### ISSUE TITLE
RATGDO - GDO, GDC Sensor Behaviour 

### ISSUE THREAD

**Issue by rcblackwell** [ratgdo/esphome-ratgdo#167]
I recently installed a RATGDO device on a Liftmaster garage door opener. Installation was straight forward and easily achieved. Once installed, I modified one of my NodeRed flows to utilize the “binary_sensor.garage_overhead_door” expecting the sensor to report on when the door was open or closed. Unfortunately the flow failed to work properly as the “binary_sensor.garage_overhead_door” sensor doesn’t report a steady state. It flips between “door was opened” and “door was closed” at approximately 2 second intervals.

The RATGDO is flashed with ESPHOME. Here’s a copy of my Home Assistant log during the time the garage door was opened then closed;

```
Garage - Overhead Door was closed
9:51:39 AM - 3 hours ago
Garage - Overhead Door was opened
9:51:32 AM - 3 hours ago
Garage - Overhead Door was closed
9:51:32 AM - 3 hours ago
Garage - Overhead Door was opened
9:51:32 AM - 3 hours ago
Garage - Overhead Door was closed
9:51:32 AM - 3 hours ago
Garage - Overhead Door was opened
9:51:31 AM - 3 hours ago
Garage - Overhead Door was closed
9:51:31 AM - 3 hours ago
Garage Door Motor turned off
9:51:28 AM - 3 hours ago
Garage Door Door was opened
9:51:28 AM - 3 hours ago
Garage - Overhead Door was opened
9:51:25 AM - 3 hours ago
Garage - Overhead Door was closed
9:51:25 AM - 3 hours ago
Garage - Overhead Door was opened
9:51:23 AM - 3 hours ago
Garage - Overhead Door was closed
9:51:23 AM - 3 hours ago
Garage - Overhead Door was opened
9:51:22 AM - 3 hours ago
Garage - Overhead Door was closed
9:51:22 AM - 3 hours ago
Garage - Overhead Door was opened
9:51:19 AM - 3 hours ago
Garage - Overhead Door was closed
9:51:19 AM - 3 hours ago
Garage - Overhead Door was opened
9:51:18 AM - 3 hours ago
Garage - Overhead Door was closed
9:51:18 AM - 3 hours ago
Garage - Overhead Door was opened
9:51:17 AM - 3 hours ago
Garage - Overhead Door was closed
9:51:17 AM - 3 hours ago
Garage Door Motor turned on
9:51:16 AM - 3 hours ago
Garage Door Door is opening
9:51:15 AM - 3 hours ago
Garage Door Light turned on
9:51:15 AM - 3 hours ago
Garage - Overhead Door was opened
9:51:15 AM - 3 hours ago
Garage Door Button turned off
9:51:14 AM - 3 hours ago
Garage Door Button turned on
9:51:14 AM - 3 hours ago
```
Any idea what could be causing this behaviour?

**Comment by rcblackwell** [comment]
Here's a little more information related to my installation and experience;

Liftmaster garage door opener - Model 8355W
Liftmaster GD remote control - Model 882LMW

Both manufactured in 2016.

While troubleshooting the issue I disconnected the remote control to see if that might stop repeated messages when the door was open. It did not.

**Comment by rcblackwell** [comment]
Stumbled across something interesting that I thought I should share. If the RATGDO logs are viewed using the ESPHome Log viewer rather than the Home Assistant log viewer, the multiple open, close entries that are present in the latter log don't appear. Here's a copy of the log as viewed directly from the ESPHome Log;

```
INFO ESPHome 2023.12.6
INFO Reading configuration /config/esphome/garage-door-opener.yaml...
INFO Updating https://github.com/ratgdo/esphome-ratgdo@None
INFO Updating https://github.com/ratgdo/esphome-ratgdo@None
INFO Detected timezone 'America/Toronto'
INFO Starting log output from 192.168.86.82 using esphome API
INFO Successfully connected to garage-door-opener @ 192.168.86.82 in 0.005s
INFO Successful handshake with garage-door-opener @ 192.168.86.82 in 2.559s
[13:42:28][I][app:102]: ESPHome version 2023.12.6 compiled on Jan 15 2024, 12:50:45
[13:42:28][I][app:104]: Project ratgdo.esphome version 2.5i
[13:42:28][C][wifi:573]: WiFi:
[13:42:28][C][wifi:405]:   Local MAC: 08:3A:8D:F9:E0:C6
[13:42:28][C][wifi:410]:   SSID: [redacted]
[13:42:28][C][wifi:411]:   IP Address: 192.168.86.82
[13:42:28][C][wifi:412]:   BSSID: [redacted]
[13:42:28][C][wifi:414]:   Hostname: 'garage-door-opener'
[13:42:28][C][wifi:416]:   Signal strength: -48 dB ▂▄▆█
[13:42:28][C][wifi:420]:   Channel: 11
[13:42:28][C][wifi:421]:   Subnet: 255.255.255.0
[13:42:28][C][wifi:422]:   Gateway: 192.168.86.1
[13:42:28][C][wifi:423]:   DNS1: 192.168.86.1
[13:42:28][C][wifi:424]:   DNS2: 0.0.0.0
[13:42:28][C][logger:439]: Logger:
[13:42:28][C][logger:440]:   Level: DEBUG
[13:42:28][C][logger:441]:   Log Baud Rate: 115200
[13:42:28][C][logger:443]:   Hardware UART: UART0
[13:42:28][C][ratgdo.number:021]: RATGDO Number 'Rolling code counter'
[13:42:28][C][ratgdo.number:021]:   Unit of Measurement: 'codes'
[13:42:28][C][ratgdo.number:025]:   Type: Rolling Code Counter
[13:42:28][C][ratgdo.number:021]: RATGDO Number 'Opening duration'
[13:42:28][C][ratgdo.number:021]:   Unit of Measurement: 's'
[13:42:28][C][ratgdo.number:027]:   Type: Opening Duration
[13:42:28][C][ratgdo.number:021]: RATGDO Number 'Closing duration'
[13:42:28][C][ratgdo.number:021]:   Unit of Measurement: 's'
[13:42:28][C][ratgdo.number:029]:   Type: Closing Duration
[13:42:28][C][ratgdo.number:021]: RATGDO Number 'Client ID'
[13:42:28][C][ratgdo.number:023]:  Type: Client ID
[13:42:28][C][switch.gpio:068]: GPIO Switch 'Status door'
[13:42:28][C][switch.gpio:090]:   Restore Mode: always OFF
[13:42:28][C][switch.gpio:031]:   Pin: GPIO16
[13:42:28][C][switch.gpio:068]: GPIO Switch 'Status obstruction'
[13:42:28][C][switch.gpio:090]:   Restore Mode: always OFF
[13:42:28][C][switch.gpio:031]:   Pin: GPIO15
[13:42:28][C][gpio.binary_sensor:015]: GPIO Binary Sensor 'Dry contact open'
[13:42:28][C][gpio.binary_sensor:016]:   Pin: GPIO14
[13:42:28][C][gpio.binary_sensor:015]: GPIO Binary Sensor 'Dry contact close'
[13:42:28][C][gpio.binary_sensor:016]:   Pin: GPIO12
[13:42:28][C][gpio.binary_sensor:015]: GPIO Binary Sensor 'Dry contact light'
[13:42:28][C][gpio.binary_sensor:016]:   Pin: GPIO0
[13:42:28][C][light:103]: Light 'Light'
[13:42:28][C][ratgdo:081]: Setting up RATGDO...
[13:42:28][C][ratgdo:082]:   Output GDO Pin: GPIO5
[13:42:28][C][ratgdo:083]:   Input GDO Pin: GPIO4
[13:42:28][C][ratgdo:087]:   Input Obstruction Pin: GPIO13
[13:42:28][C][ratgdo:089]:   Rolling Code Counter: 94
[13:42:28][C][ratgdo:090]:   Client ID: 7488825
[13:42:28][C][ratgdo.sensor:012]: RATGDO Sensor 'Openings'
[13:42:28][C][ratgdo.sensor:012]:   State Class: ''
[13:42:28][C][ratgdo.sensor:012]:   Unit of Measurement: 'openings'
[13:42:28][C][ratgdo.sensor:012]:   Accuracy Decimals: 0
[13:42:28][C][ratgdo.sensor:012]:   Icon: 'mdi:open-in-app'
[13:42:28][C][ratgdo.sensor:013]:   Type: Openings
[13:42:28][C][ratgdo.lock:012]: RATGDO Lock 'Lock remotes'
[13:42:28][C][ratgdo.lock:013]:   Type: Lock
[13:42:28][C][ratgdo.binary_sensor:036]: RATGDO BinarySensor 'Motion'
[13:42:28][C][ratgdo.binary_sensor:036]:   Device Class: 'motion'
[13:42:28][C][ratgdo.binary_sensor:038]:   Type: Motion
[13:42:28][C][ratgdo.binary_sensor:036]: RATGDO BinarySensor 'Obstruction'
[13:42:28][C][ratgdo.binary_sensor:036]:   Device Class: 'problem'
[13:42:28][C][ratgdo.binary_sensor:040]:   Type: Obstruction
[13:42:28][C][ratgdo.binary_sensor:036]: RATGDO BinarySensor 'Button'
[13:42:28][C][ratgdo.binary_sensor:044]:   Type: Button
[13:42:28][C][ratgdo.binary_sensor:036]: RATGDO BinarySensor 'Motor'
[13:42:28][C][ratgdo.binary_sensor:036]:   Device Class: 'running'
[13:42:28][C][ratgdo.binary_sensor:042]:   Type: Motor
[13:42:28][C][ratgdo.cover:014]: RATGDO Cover 'Door'
[13:42:28][C][ratgdo.cover:014]:   Device Class: 'garage'
[13:42:28][C][ratgdo.light:014]: RATGDO Light
[13:42:28][C][restart.button:017]: Restart Button 'Restart'
[13:42:28][C][safe_mode.button:022]: Safe Mode Button 'Safe mode boot'
[13:42:28][C][safe_mode.button:022]:   Icon: 'mdi:restart-alert'
[13:42:28][C][homeassistant.time:010]: Home Assistant Time:
[13:42:28][C][homeassistant.time:011]:   Timezone: 'EST5EDT,M3.2.0,M11.1.0'
[13:42:28][C][web_server:168]: Web Server:
[13:42:28][C][web_server:169]:   Address: garage-door-opener.local:80
[13:42:28][C][mdns:115]: mDNS:
[13:42:28][C][mdns:116]:   Hostname: garage-door-opener
[13:42:28][C][ota:097]: Over-The-Air Updates:
[13:42:28][C][ota:098]:   Address: garage-door-opener.local:8266
[13:42:28][C][api:139]: API Server:
[13:42:28][C][api:140]:   Address: garage-door-opener.local:6053
[13:42:28][C][api:142]:   Using noise encryption: YES
[13:42:28][C][improv_serial:032]: Improv Serial:
[13:42:35][D][cover:076]: 'Door' - Setting
[13:42:35][D][cover:084]:   Position: 100%
[13:42:35][D][number:012]: 'Rolling code counter': Sending state 95.000000
[13:42:36][D][ratgdo:210]: Status: door=OPENING light=ON lock=UNLOCKED
[13:42:36][D][cover:170]: 'Door' - Publishing:
[13:42:36][D][cover:173]:   Position: 0%
[13:42:36][D][cover:186]:   Current Operation: OPENING
[13:42:36][D][ratgdo:228]: Motor: state=ON
[13:42:36][D][binary_sensor:036]: 'Motor': Sending state ON
[13:42:36][D][esp8266.preferences:238]: Saving preferences to flash...
[13:42:37][D][cover:170]: 'Door' - Publishing:
[13:42:37][D][cover:173]:   Position: 4%
[13:42:37][D][cover:186]:   Current Operation: OPENING
[13:42:37][D][cover:170]: 'Door' - Publishing:
[13:42:37][D][cover:173]:   Position: 8%
[13:42:37][D][cover:186]:   Current Operation: OPENING
[13:42:38][D][cover:170]: 'Door' - Publishing:
[13:42:38][D][cover:173]:   Position: 12%
[13:42:38][D][cover:186]:   Current Operation: OPENING
[13:42:38][D][cover:170]: 'Door' - Publishing:
[13:42:38][D][cover:173]:   Position: 16%
[13:42:38][D][cover:186]:   Current Operation: OPENING
[13:42:39][D][cover:170]: 'Door' - Publishing:
[13:42:39][D][cover:173]:   Position: 20%
[13:42:39][D][cover:186]:   Current Operation: OPENING
[13:42:39][D][cover:170]: 'Door' - Publishing:
[13:42:39][D][cover:173]:   Position: 23%
[13:42:39][D][cover:186]:   Current Operation: OPENING
[13:42:40][D][cover:170]: 'Door' - Publishing:
[13:42:40][D][cover:173]:   Position: 27%
[13:42:40][D][cover:186]:   Current Operation: OPENING
[13:42:40][D][cover:170]: 'Door' - Publishing:
[13:42:40][D][cover:173]:   Position: 31%
[13:42:40][D][cover:186]:   Current Operation: OPENING
[13:42:41][D][cover:170]: 'Door' - Publishing:
[13:42:41][D][cover:173]:   Position: 35%
[13:42:41][D][cover:186]:   Current Operation: OPENING
[13:42:41][D][cover:170]: 'Door' - Publishing:
[13:42:41][D][cover:173]:   Position: 39%
[13:42:41][D][cover:186]:   Current Operation: OPENING
[13:42:41][D][ratgdo:228]: Motor: state=ON
[13:42:41][D][esp8266.preferences:238]: Saving preferences to flash...
[13:42:42][D][cover:170]: 'Door' - Publishing:
[13:42:42][D][cover:173]:   Position: 43%
[13:42:42][D][cover:186]:   Current Operation: OPENING
[13:42:42][D][cover:170]: 'Door' - Publishing:
[13:42:42][D][cover:173]:   Position: 47%
[13:42:42][D][cover:186]:   Current Operation: OPENING
[13:42:43][D][cover:170]: 'Door' - Publishing:
[13:42:43][D][cover:173]:   Position: 51%
[13:42:43][D][cover:186]:   Current Operation: OPENING
[13:42:43][D][ratgdo:228]: Motor: state=ON
[13:42:43][D][cover:170]: 'Door' - Publishing:
[13:42:43][D][cover:173]:   Position: 55%
[13:42:43][D][cover:186]:   Current Operation: OPENING
[13:42:44][D][cover:170]: 'Door' - Publishing:
[13:42:44][D][cover:173]:   Position: 59%
[13:42:44][D][cover:186]:   Current Operation: OPENING
[13:42:44][D][cover:170]: 'Door' - Publishing:
[13:42:44][D][cover:173]:   Position: 63%
[13:42:44][D][cover:186]:   Current Operation: OPENING
[13:42:45][D][cover:170]: 'Door' - Publishing:
[13:42:45][D][cover:173]:   Position: 66%
[13:42:45][D][cover:186]:   Current Operation: OPENING
[13:42:45][D][cover:170]: 'Door' - Publishing:
[13:42:45][D][cover:173]:   Position: 70%
[13:42:45][D][cover:186]:   Current Operation: OPENING
[13:42:46][D][cover:170]: 'Door' - Publishing:
[13:42:46][D][cover:173]:   Position: 74%
[13:42:46][D][cover:186]:   Current Operation: OPENING
[13:42:46][D][ratgdo:228]: Motor: state=ON
[13:42:46][D][cover:170]: 'Door' - Publishing:
[13:42:46][D][cover:173]:   Position: 78%
[13:42:46][D][cover:186]:   Current Operation: OPENING
[13:42:46][D][esp8266.preferences:238]: Saving preferences to flash...
[13:42:47][D][cover:170]: 'Door' - Publishing:
[13:42:47][D][cover:173]:   Position: 82%
[13:42:47][D][cover:186]:   Current Operation: OPENING
[13:42:47][D][cover:170]: 'Door' - Publishing:
[13:42:47][D][cover:173]:   Position: 86%
[13:42:47][D][cover:186]:   Current Operation: OPENING
[13:42:48][D][cover:170]: 'Door' - Publishing:
[13:42:48][D][cover:173]:   Position: 90%
[13:42:48][D][cover:186]:   Current Operation: OPENING
[13:42:48][D][cover:170]: 'Door' - Publishing:
[13:42:48][D][cover:173]:   Position: 94%
[13:42:48][D][cover:186]:   Current Operation: OPENING
[13:42:49][D][ratgdo:636]: Cancelling position callbacks
[13:42:49][D][ratgdo:210]: Status: door=OPEN light=ON lock=UNLOCKED
[13:42:49][D][cover:170]: 'Door' - Publishing:
[13:42:49][D][cover:173]:   Position: 100%
[13:42:49][D][cover:186]:   Current Operation: IDLE
[13:42:49][D][switch:012]: 'Status door' Turning ON.
[13:42:49][D][switch:055]: 'Status door': Sending state ON
[13:42:49][D][binary_sensor:036]: 'Motor': Sending state OFF
[13:42:51][D][esp8266.preferences:238]: Saving preferences to flash...
[13:43:56][D][cover:076]: 'Door' - Setting
[13:43:56][D][cover:084]:   Position: 0%
[13:43:56][D][number:012]: 'Rolling code counter': Sending state 96.000000
[13:43:56][D][esp8266.preferences:238]: Saving preferences to flash...
[13:43:57][D][ratgdo:210]: Status: door=CLOSING light=ON lock=UNLOCKED
[13:43:57][D][cover:170]: 'Door' - Publishing:
[13:43:57][D][cover:173]:   Position: 100%
[13:43:57][D][cover:186]:   Current Operation: CLOSING
[13:43:58][D][cover:170]: 'Door' - Publishing:
[13:43:58][D][cover:173]:   Position: 96%
[13:43:58][D][cover:186]:   Current Operation: CLOSING
[13:43:58][D][ratgdo:228]: Motor: state=ON
[13:43:58][D][binary_sensor:036]: 'Motor': Sending state ON
[13:43:58][D][cover:170]: 'Door' - Publishing:
[13:43:58][D][cover:173]:   Position: 92%
[13:43:58][D][cover:186]:   Current Operation: CLOSING
[13:43:59][D][cover:170]: 'Door' - Publishing:
[13:43:59][D][cover:173]:   Position: 88%
[13:43:59][D][cover:186]:   Current Operation: CLOSING
[13:43:59][D][cover:170]: 'Door' - Publishing:
[13:43:59][D][cover:173]:   Position: 84%
[13:43:59][D][cover:186]:   Current Operation: CLOSING
[13:44:00][D][cover:170]: 'Door' - Publishing:
[13:44:00][D][cover:173]:   Position: 80%
[13:44:00][D][cover:186]:   Current Operation: CLOSING
[13:44:00][D][cover:170]: 'Door' - Publishing:
[13:44:00][D][cover:173]:   Position: 76%
[13:44:00][D][cover:186]:   Current Operation: CLOSING
[13:44:01][D][cover:170]: 'Door' - Publishing:
[13:44:01][D][cover:173]:   Position: 72%
[13:44:01][D][cover:186]:   Current Operation: CLOSING
[13:44:01][D][ratgdo:228]: Motor: state=ON
[13:44:01][D][cover:170]: 'Door' - Publishing:
[13:44:01][D][cover:173]:   Position: 68%
[13:44:01][D][cover:186]:   Current Operation: CLOSING
[13:44:01][D][esp8266.preferences:238]: Saving preferences to flash...
[13:44:02][D][cover:170]: 'Door' - Publishing:
[13:44:02][D][cover:173]:   Position: 64%
[13:44:02][D][cover:186]:   Current Operation: CLOSING
[13:44:02][D][cover:170]: 'Door' - Publishing:
[13:44:02][D][cover:173]:   Position: 60%
[13:44:02][D][cover:186]:   Current Operation: CLOSING
[13:44:03][D][cover:170]: 'Door' - Publishing:
[13:44:03][D][cover:173]:   Position: 56%
[13:44:03][D][cover:186]:   Current Operation: CLOSING
[13:44:03][D][ratgdo:228]: Motor: state=ON
[13:44:03][D][cover:170]: 'Door' - Publishing:
[13:44:03][D][cover:173]:   Position: 52%
[13:44:03][D][cover:186]:   Current Operation: CLOSING
[13:44:04][D][cover:170]: 'Door' - Publishing:
[13:44:04][D][cover:173]:   Position: 48%
[13:44:04][D][cover:186]:   Current Operation: CLOSING
[13:44:04][D][cover:170]: 'Door' - Publishing:
[13:44:04][D][cover:173]:   Position: 44%
[13:44:04][D][cover:186]:   Current Operation: CLOSING
[13:44:05][D][cover:170]: 'Door' - Publishing:
[13:44:05][D][cover:173]:   Position: 40%
[13:44:05][D][cover:186]:   Current Operation: CLOSING
[13:44:05][D][cover:170]: 'Door' - Publishing:
[13:44:05][D][cover:173]:   Position: 36%
[13:44:05][D][cover:186]:   Current Operation: CLOSING
[13:44:06][D][cover:170]: 'Door' - Publishing:
[13:44:06][D][cover:173]:   Position: 32%
[13:44:06][D][cover:186]:   Current Operation: CLOSING
[13:44:06][D][cover:170]: 'Door' - Publishing:
[13:44:06][D][cover:173]:   Position: 29%
[13:44:06][D][cover:186]:   Current Operation: CLOSING
[13:44:06][D][esp8266.preferences:238]: Saving preferences to flash...
[13:44:07][D][cover:170]: 'Door' - Publishing:
[13:44:07][D][cover:173]:   Position: 25%
[13:44:07][D][cover:186]:   Current Operation: CLOSING
[13:44:07][D][cover:170]: 'Door' - Publishing:
[13:44:07][D][cover:173]:   Position: 21%
[13:44:07][D][cover:186]:   Current Operation: CLOSING
[13:44:07][D][ratgdo:228]: Motor: state=ON
[13:44:08][D][cover:170]: 'Door' - Publishing:
[13:44:08][D][cover:173]:   Position: 17%
[13:44:08][D][cover:186]:   Current Operation: CLOSING
[13:44:08][D][cover:170]: 'Door' - Publishing:
[13:44:08][D][cover:173]:   Position: 13%
[13:44:08][D][cover:186]:   Current Operation: CLOSING
[13:44:09][D][cover:170]: 'Door' - Publishing:
[13:44:09][D][cover:173]:   Position: 9%
[13:44:09][D][cover:186]:   Current Operation: CLOSING
[13:44:09][D][cover:170]: 'Door' - Publishing:
[13:44:09][D][cover:173]:   Position: 5%
[13:44:09][D][cover:186]:   Current Operation: CLOSING
[13:44:09][D][ratgdo:636]: Cancelling position callbacks
[13:44:09][D][ratgdo:210]: Status: door=CLOSED light=ON lock=UNLOCKED
[13:44:09][D][cover:170]: 'Door' - Publishing:
[13:44:09][D][cover:173]:   Position: 0%
[13:44:09][D][cover:186]:   Current Operation: IDLE
[13:44:09][D][switch:016]: 'Status door' Turning OFF.
[13:44:09][D][switch:055]: 'Status door': Sending state OFF
[13:44:09][D][number:012]: 'Rolling code counter': Sending state 97.000000
[13:44:09][D][binary_sensor:036]: 'Motor': Sending state OFF
[13:44:10][D][ratgdo:237]: Openings: 4067
[13:44:10][D][sensor:093]: 'Openings': Sending state 4067.00000 openings with 0 decimals of accuracy
[13:44:11][D][esp8266.preferences:238]: Saving preferences to flash...
```

**Comment by PaulWieland** [comment]
have you upgraded home assistant to the latest release?

**Comment by rcblackwell** [comment]
I'm running;

```
Core - 2024.1.3
Supervisor - 2023.12.0
Operating System - 11.4
Frontend - 20240104.0
```

on an HA Blue (ODROID-N2/N2+)

**Comment by rcblackwell** [comment]
I found the problem and, boy, do I ever feel **REALLY, REALLY STUPID!**

I have an ESP8266 programmed and attached to reed switches that detect the state of my exterior doors. One of those doors is, you guest it, the garage door. When setting up a NodeRed flow for exterior lighting triggered by the garage door opening, I mistakenly selected the reed switch named Garage Door. Things started acting up again today so I dug deeply into the logs. It took some time to realize the error I made. With the correct sensor now selected, my NodeRed lighting flow is working the way it should.

Many apologies for the misleading information. I hope I haven't caused to much trouble for you.
